### ISSUE TITLE
Text 2 Subtitle

### ISSUE THREAD

**Issue by tenshiism** [aedocw/epub2tts#213]
There is no way to pass a param to create a .srt subtitle file when speaking from a book.
its sometimes hard to read along with a book while listening, so it would be good to be able to read the original text and not have "harvard" be rendered as "hamburger" when generating a whisper file later. 
Where is the method that checks the TTS generation against the original text with whisper? If you can tell me, I can maybe use it to make this feature.

**Comment by aedocw** [comment]
Line 314 in epub2tts.py, this function:

```
    def compare(self, text, wavfile):
        result = self.whispermodel.transcribe(wavfile)
        text = re.sub(" +", " ", text).lower().strip()
        ratio = fuzz.ratio(text, result["text"].lower())
        print(f"Transcript: {result['text'].lower()}") if self.debug else None
        print(
            f"Text to transcript comparison ratio: {ratio}"
        ) if self.debug else None
        return ratio
```

I'm confused as to what you are talking about though, why would you need an srt file if you have the original book in either epub or txt format?

**Comment by tenshiism** [comment]
tl;dr below but simple
Use case:
After making voice with epub2tts, you want to read along, but the epub is a pain to read. So you render the mp3 to mp4 and subtitle it with whisper. But the subs are wrong. So you try to output srt on epub2tts - but theres no option for that.

I'm confused that you don't understand. Do you read your audiobooks while you listen to them? I thought most people learnt to read these days by watching books on youtube...
This image illustrates the point.
![image](https://github.com/aedocw/epub2tts/assets/143989078/5a531051-5cfd-41b5-8f9b-f6edc1c8a058)
Every time it says the characters name, it uses a different spelling, due to whisper running a second time. in order to generate the video  subtitles, as it stands you have to run it through whisper with a script, or another a subtitle program that uses whisper like "subtitle edit" - which often takes longer than making the speech in the first place with epub2tts

Since epub2tts can do this, it seems a waste not to.

**Comment by aedocw** [comment]
Ah, I understand this use-case. Unfortunately it would not be possible for epub2tts to create an srt file while creating the original audiobook, because the timing would quickly become incorrect - at the end of the creation of each chapter, all silences longer than 1 second are removed from the audio stream. This removes inevitable long pauses, but it means when several (sometimes dozens) of seconds are removed throughout the book, so the drift of incorrect timing as far as whisper is concerned would get worse as worse as the book goes on.

The other issue is the way whisper is used in this case - whisper is run against each individual sentence, so whisper doesn't have any idea of the overall timing, thus it could not create a full SRT file.

You might be interested in looking at https://smoores.gitlab.io/storyteller/ which does something similar, highlighting sections of the book while listening to an audiobook by creating an epub3 version of the book.

The other thing you might check out is https://github.com/Vaibhavs10/insanely-fast-whisper which can be MUCH faster than whisper.

**Comment by tenshiism** [comment]
> The other issue is the way whisper is used in this case - whisper is run against each individual sentence, so whisper doesn't have any idea of the overall timing, thus it could not create a full SRT file.

Whisper? Whisper is what is used to create srt files in programs like "subtitle edit" or https://github.com/abdeladim-s/subsai .. 
But isn't the audio created with RVC in 1-20s batches? One could probably pair it to the segments befure theyre fused... It doesn't seem hard but maybe it is...

**Comment by tenshiism** [comment]
By the way, are you aware that about 5% of the text from the epub is not present in the final file? at least when using whisper ratios of 2-4 (because 1 is full of groaning voice sounds)

**Comment by aedocw** [comment]
I will not be able to put any time into implementing a feature like this, but if you would like to submit a PR I would be happy to test and review it.
